### DISCUSSION TITLE
Welcome to quicksnap Discussions!

### DISCUSSION THREAD

**Discussion by JulienHeijmans** [JulienHeijmans/quicksnap#2] [Ideas]
## 👋 Welcome!
  If you have suggestions, feature ideas, you are welcome to share them below in the comment area.

**Comment by etma101** [comment]
hi, your addon is amazing. i have a feature idea: At current version, when selecting the first vertex we need to hold mouse down and release at destination vertex. it would be cool to have a setting for allowing mouse release: Clicking on first vertex then clicking at target vertex for confirm. also in between first and second click maybe we can continue to use the middle mouse button for rotating and panning the view for better angles. I saw this feature in Sketchup and its very useful. thank you again for this amazing tool.

**Reply by JulienHeijmans** [reply]
Hey, thank you for the suggestion!

I totally see the utility of this (Sapping to points that are outside of the camera view when you pick the origin point of the snap).
I will definitely add it to my list of potential updates.

Just a question, how do you think it should behave after you have picked the origin point with a click... I can see different possibilities:

-  The selected object/points could be constantly following the mouse like right now (Including when you move the camera), Including the rubber band.
(I think it could look a bit weird.)
- The selected object could be moved, and the rubber band displayed, only when there is a target point under the mouse.
(Might be a bit strange when you move the mouse between two snap point, the object would dissapear, then reappear when the new snap target is found)
- Same as above, but when the user press and hold, then the object will stick under the mouse.
- Maybe you have some other suggestion ? How does it work  in Sketchup? 


Or maybe only when the user hold the

**Reply by etma101** [reply]
i think seeing the object while moving it is essential for visualising the final look before confirm. Sketchup is similar to your first scenario. object sticks to mouse after clicking and a rubberband is visible.

https://user-images.githubusercontent.com/115644588/195396299-96022a4f-ae73-4b87-bd04-4180999831dd.mov

.

**Reply by etma101** [reply]
Although it might be hard to implement in blender since it has no view control on other commands like generic move, scale, rotate etc.

**Reply by JulienHeijmans** [reply]
Aaah yes ok, so the object is actually static while the camera moves, and the snapping start again once the movement is finishes.
Alright, noted. It won't come right now, but I will look at it when I have time. It should not be hard to implement actually, I am already doing something similar before the snapping starts (I let the middle mouse event pass through and it's picked up by blender default camera control behaviour)

Thank you for the suggestion !
I will post a reply here when I have it working :)

**Reply by JulienHeijmans** [reply]
Hey Etma101, 
I wanted to let you know that [the feature you reqested is now in the tool](https://github.com/JulienHeijmans/quicksnap/discussions/25), it is now possible the snap either by click&drag or by clicking once to pick the snap origin, and clicking a second time to pick the snap destination. And it is possible to move the camera in between the two. 
You can get it by simply updating the addon in the addon preferences.

Thank you for making the suggestion and helping make the tool better!

**Reply by etma101** [reply]
honestly i thought it was impossible to adapt this function to blender but it works flawless. you did an amazing job! thank you

**Comment by Aritaborian** [comment]
Hello and thanks for making such an amazing tool, I've already shared it with couple of Russian-spoken Blender-devoted communities in Telegram. Let me share some ideas.
Currently, snapping is limited to the object's movement only. The ability to use snapping during scaling and rotation would greatly expand the possibilities.
Further, the possibilities of snapping between different types of elements are interesting: vertex-to-edge, edge-to-face, etc.
As an example of usability of these ideas imagine the situation when we rotate one object around given axis until its active (selected) vertex touches some edge or face of another object.

**Reply by JulienHeijmans** [reply]
Hi! And thank you for the suggestion, and for sharing the tool around ! :)

> Further, the possibilities of snapping between different types of elements are interesting: vertex-to-edge, edge-to-face, etc.

This one, or at least some form of that is in development, if you want to try, you can check it out here: [](https://github.com/JulienHeijmans/quicksnap/tree/midedge-facecenter-snapping) 
The way it works is that you start the tool, then when the tool is running, you press the same hotkey as for starting hte tool, it will show a menu that lets you chose if you want to snap from/to:
- vertices, curve points, etc
- edge midpoints
- face centers

>  The ability to use snapping during scaling and rotation would greatly expand the possibilities.

For the moment, I was focusing on building a simple tool that fits the most common cases, I think that allowing the tool to work during rotation/scale might be an action that might be more of a rare occasion (~one every few days), while simple snapping is an action I see as a very common case (many times in a day), it would also drastically increase the complexity of the tool (On my side), as the user would probably want to pick the rotation/scale center first, and then the snap origin point and then the target point. THat would involve quite some refactoring.
I will still add that on my list of potential future feature, and maybe if more people seem to be interested I might add it to the addon. 
But there are other feature that I see as more relevant to do in the first place.

Note: If someone is reading this message also thinks this is a feature you would really like, please click the arrow up on the feature request message from Aritaborian above :)

**Reply by Aritaborian** [reply]
Thanks for paying attention.
> This one, or at least some form of that is in development, if you want to try, you can check it out here:

Where? Probably, smth is missing.

**Reply by JulienHeijmans** [reply]
Sorry, it looks like I forgot to add the link.

But I pushed the changes on main, so if you update the addon (in the addon preferences) you should get it. 
I updated the explaination [in the readme](https://github.com/JulienHeijmans/quicksnap), showing how do align from/to edge midpoints and face centers.
